### ISSUE TITLE
Consolidate and document order between cancellation handlers and continuation resume

### ISSUE THREAD

**Issue by qwwdfsad** [Kotlin/kotlinx.coroutines#415]
Reproducer from #407 will fail with exception in coroutines machinery

```
@Test
fun testCancelProduce() = runTest {
    val source = Channel<Int>()
    val produced = produce<Int>(ctx, onCompletion = source.consumes()) {
        source.receive() // <- when exception is thrown from here, source is NOT empty for Unconfined
    }
  
    produced.cancel()
}
```
Behaviour will be different for `Unconfined` and `coroutineContext` as `ctx`.

It's because we do the following:
```
cont.cancel(e) 
->  dispatch() { cont.resumeWithException(e) }
->  for (handler in handler) handler.invoke(e)
```

We have data-race with small window between dispatched continuation and handlers invocation. Same race is 100% reproducible with Unconfined.

**Comment by PaulWoitaschek** [comment]
Can this please be released soon? About 0.7% of my sessions are crashing with this exception.

**Comment by qwwdfsad** [comment]
Presumably it will be released next week
